File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        memo = defaultdict(int)
        def dp(x,y):
            if x + y <= 1: return 1
            state = (x,y)
            if state not in memo:
                if x >= 1:memo[state] += dp(x-1,y)
                if y >= 1: memo[state] += dp(x,y-1)
            return memo[state]
        return dp(n-1,m-1)
                
        

File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        if(n == 1 || m == 1) return 1;
        int x = m + n - 2;
        int y = min(n-1,m-1);
        unsigned long long ans = 1;
        for (int i=1; i<=y; i++) ans = ans*(x-y+i)/i; 
        return ans;
        
    }
};